### PR TITLE
Initial Kokkos implementation of NemoLite2D

### PR DESCRIPTION
**Pull request by sergisiso**
The initial implementation has the kokkos parallel dispatch and the Execution Policies (for the host), but it doesn't include targeting acceleration devices or Views (which are necessary for performance portability).

For the host it matches the C++ OpenMP (which it uses under-the-hood), so it doesn't introduce any overhead at least for the simple implementation.
```
Implementation          Compiler        Arch    ua checksum     uv checksum     time/step
psykal_serial:          gcc             cpu     0.41022150E-01  0.50252378E+01  0.15323E+00
psykal_cpp:             gcc             cpu     0.41022150E-01  0.50252378E+01  0.11709E+00
psykal_omp:             gcc             cpu     0.41022150E-01  0.50252378E+01  0.35053E-01
psykal_cpp_omp:         gcc             cpu     0.41022150E-01  0.50252378E+01  0.29249E-01
psykal_kokkos:          gcc             cpu     0.41022150E-01  0.50252378E+01  0.29679E-01
```

### CODE REVIEW

**Comment by sergisiso** [comment]
@arporter @rupertford The initial (host only / no data layout abstraction) kokkos version is ready for review. It is conceptually quite similar to the C++ OpenMP version with only a different syntax. When writing the README of this implementation I realized I didn't add one for the `psykal_cpp` , so it is also part of this PR.

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# C++ PSyKal versions of NEMOLite2D #
+
+This directory contains various version of the PSyKAl form of
+NEMOLite2D that has had its PSy layers manually implemented in C++.
+
+The Fortran(NemoLite2d) to C++(PSy-layer) interoperability is done
+with a `iso_c_binding` interface of the function `c_invoke_time_step`
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/README.md:7]: s/a/an/ and s/of the/for the/

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/Makefile [hunk 2]
```diff
@@ -0,0 +1,69 @@
+# Makefile for manual C++ kokkos version of PSyKAl'd NEMOLite2D.
+#
+# This Makefile picks up the compiler to use any flags from
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/Makefile:3]: s/to use any/to use plus any/

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/README.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+is done with the interface introduced in the `psykal_cpp` version.
+
+Kokkos is a Performance Portability EcoSystem for C++,
+a good introduction to its concepts and APIs can be found in their
+(wiki)[https://github.com/kokkos/kokkos/wiki].
+
+
+## Implementation limitations ##
+
+Kokkos has several abstraction to control where (execution spaces),
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/README.md:15]: s/abstraction/abstractions/ and s/,/and/

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        KOKKOS_LAMBDA (const int jj, const int ji) {
+            int idx = jj * width + ji;
+            un[idx] = ua[idx];
+            vn[idx] = va[idx];
+            sshn_t[idx] = ssha_t[idx];
+        }
+    );
+
+    // Time update kernel (internal domain u points)
+    Kokkos::parallel_for("continuity",
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp:169]: s/continuity/next_sshu/ I think?

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        mdrange_policy({internal_ystart, internal_xstart},
+                       {internal_ystop, internal_xstop - 1}),
+        KOKKOS_LAMBDA (const int jj, const int ji) {
+            next_sshu_code(ji, jj, width, sshn_u, sshn_t, tmask, area_t, area_u);
+        }
+    );
+
+    // Time update kernel (internal domain v points)
+    Kokkos::parallel_for("continuity",
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp:178]: s/continuity/next_sshv/ I think?

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    Kokkos::parallel_for("bc_solid_v",
+        mdrange_policy({internal_ystart - 1, internal_xstart - 1},
+                       {internal_ystop, internal_xstop + 1}),
+        KOKKOS_LAMBDA (const int jj, const int ji) {
+            bc_flather_v_code(ji, jj, width, va, hv, sshn_v, tmask, g);
+        }
+    );
+
+    // Copy 'next' fields to 'current' fields (whole domain)
+    Kokkos::parallel_for("bc_solid_v",
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp:157]: s/bc_solid_v/copy_fields/ or similar?

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/README.md (hunk 1 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_cpp/README.md:7]: Done

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/Makefile (hunk 2 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/Makefile:3]: Done

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/README.md (hunk 3 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/README.md:15]: Done

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp (hunk 4 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp:169]: Ups my copy-pasting shows here :)

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp (hunk 5 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp:178]: Fixed

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp (hunk 6 above)
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp:157]: Fixed

**Comment by sergisiso** [comment]
@arporter I fixed the typos and corrected the momentum loop boundaries of C++ and kokkos like in the upcoming PR #37 . This is ready for the next review.

File: benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp [hunk 7]
```diff
@@ -84,10 +84,10 @@ extern "C" void c_invoke_time_step(
         }
     );
 
-    // Momentum_u kernel (internal domain)
+    // Momentum_u kernel (internal domain but top x)
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp:87]: Apologies if I'm being picky but I don't understand what 'but top x' means. Perhaps you want "internal u points" and similar for v below?

Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/manual_versions/psykal_kokkos/time_step_kokkos.cpp]: Yes, that sounds better, changed.

**Comment by sergisiso** [comment]
@arporter Fixed the remaining issue, ready for next review

**Pull request merged**